File: IntelligentDriving/src/dynamics.jl
##^# rollout methods ###########################################################
# mutating way
function rollout1(U, x0, f, fx, fu, X_prev, U_prev)
  xdim, N = size(X_prev)
  x = x0
  X = zeros(xdim, N + 1)
  X[:, 1] = x
  for i = 1:N
    x =
      f[:, i] +
      fx[:, :, i] * (x - X_prev[:, i]) +
      fu[:, :, i] * (U[:, i] - U_prev[:, i])
    X[:, i + 1] = x
  end
  return X
end

# tuple way
function rollout2(U, x0, f, fx, fu, X_prev, U_prev)
  xdim, N = size(X_prev)
  x = x0
  xs = (x,)
  for i = 1:N
    x =
      f[:, i] +
      fx[:, :, i] * (x - X_prev[:, i]) +
      fu[:, :, i] * (U[:, i] - U_prev[:, i])
    xs = (xs..., x)
  end
  return reduce(hcat, xs)
end

# cat way
function rollout3(U, x0, f, fx, fu, X_prev, U_prev)
  xdim, N = size(X_prev)
  x = x0
  X = x
  for i = 1:N
    x =
      f[:, i] +
      fx[:, :, i] * (x - X_prev[:, i]) +
      fu[:, :, i] * (U[:, i] - U_prev[:, i])
    X = hcat(X, x)
  end
  return X
end

# Zygote.Buffer way
function rollout4(U, x0, f, fx, fu, X_prev, U_prev)
  xdim, N = size(X_prev)
  x = x0
  x_list = Zygote.Buffer(typeof(x0)[], N + 1)
  x_list[1] = x
  for i = 1:N
    x =
      f[:, i] +
      fx[:, :, i] * (x - X_prev[:, i]) +
      fu[:, :, i] * (U[:, i] - U_prev[:, i])
    x_list[i + 1] = x
  end
  return reduce(hcat, copy(x_list))
end

# select version 4 as the default rollout, it is the best and differentiable
rollout = rollout4
##$#############################################################################
##^# unicycle dynamics #########################################################
function unicycle_f(x, u, p)
  #=
  unicycle car dynamics, 4 states, 2 actions
  x1: position x
  x2: position y
  x3: speed (local frame)
  x4: orientation angle

  u1: acceleration
  u2: turning speed (independent of velocity)
  =#

  @assert length(x) == 4 && length(u) == 2 && length(p) == 3
  dt, u_scale = p[1], p[2:3]
  u = copy(u)
  eps = 1e-6
  u = u .+ eps * (u .>= 0.0)

  f = Zygote.Buffer(Float64[], 4)
  f[1] =
    (
      (u[1] * u_scale[1] * u[2] * u_scale[2] * dt + u[2] * u_scale[2] * x[3]) *
      sin(u[2] * u_scale[2] * dt + x[4]) +
      u[1] * u_scale[1] * cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u[2] * u_scale[2] * x[3] * sin(x[4]) + u[1] * u_scale[1] * cos(x[4])) /
    (u[2]^2 * u_scale[2]^2) + x[1]
  f[2] =
    (
      u[1] * u_scale[1] * sin(u[2] * u_scale[2] * dt + x[4]) +
      (
        (-u[1] * u_scale[1] * u[2] * u_scale[2] * dt) -
        u[2] * u_scale[2] * x[3]
      ) * cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u[1] * u_scale[1] * sin(x[4]) - u[2] * u_scale[2] * x[3] * cos(x[4])) /
    (u[2]^2 * u_scale[2]^2) + x[2]
  f[3] = u[1] * u_scale[1] * dt + x[3]
  f[4] = u[2] * u_scale[2] * dt + x[4]
  return copy(f)
end

function unicycle_fx(x, u, p)
  @assert length(x) == 4 && length(u) == 2 && length(p) == 3
  dt, u_scale = p[1], p[2:3]
  u = copy(u)
  eps = 1e-6
  u = u .+ eps * (u .>= 0.0)

  fx = zeros(4, 4)
  fx[1, 1] = 1
  fx[1, 2] = 0
  fx[1, 3] =
    sin(u[2] * u_scale[2] * dt + x[4]) / (u[2] * u_scale[2]) -
    sin(x[4]) / (u[2] * u_scale[2])
  fx[1, 4] =
    (
      (u[1] * u_scale[1] * u[2] * u_scale[2] * dt + u[2] * u_scale[2] * x[3]) *
      cos(u[2] * u_scale[2] * dt + x[4]) -
      u[1] * u_scale[1] * sin(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u[2] * u_scale[2] * x[3] * cos(x[4]) - u[1] * u_scale[1] * sin(x[4])) /
    (u[2]^2 * u_scale[2]^2)
  fx[2, 1] = 0
  fx[2, 2] = 1
  fx[2, 3] =
    cos(x[4]) / (u[2] * u_scale[2]) -
    cos(u[2] * u_scale[2] * dt + x[4]) / (u[2] * u_scale[2])
  fx[2, 4] =
    (
      u[1] * u_scale[1] * cos(u[2] * u_scale[2] * dt + x[4]) -
      (
        (-u[1] * u_scale[1] * u[2] * u_scale[2] * dt) -
        u[2] * u_scale[2] * x[3]
      ) * sin(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u[2] * u_scale[2] * x[3] * sin(x[4]) + u[1] * u_scale[1] * cos(x[4])) /
    (u[2]^2 * u_scale[2]^2)
  fx[3, 1] = 0
  fx[3, 2] = 0
  fx[3, 3] = 1
  fx[3, 4] = 0
  fx[4, 1] = 0
  fx[4, 2] = 0
  fx[4, 3] = 0
  fx[4, 4] = 1
  return fx
end

function unicycle_fu(x, u, p)
  @assert length(x) == 4 && length(u) == 2 && length(p) == 3
  dt, u_scale = p[1], p[2:3]
  u = copy(u)
  eps = 1e-6
  u = u .+ eps * (u .>= 0.0)

  fu = zeros(4, 2)
  fu[1, 1] =
    (
      u_scale[1] * u[2] * u_scale[2] * dt * sin(u[2] * u_scale[2] * dt + x[4]) +
      u_scale[1] * cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u_scale[1] * cos(x[4])) / (u[2]^2 * u_scale[2]^2)
  fu[1, 2] =
    (
      -(
        2 * (
          (
            u[1] * u_scale[1] * u[2] * u_scale[2] * dt +
            u[2] * u_scale[2] * x[3]
          ) * sin(u[2] * u_scale[2] * dt + x[4]) +
          u[1] * u_scale[1] * cos(u[2] * u_scale[2] * dt + x[4])
        )
      ) / (u[2]^3 * u_scale[2]^2)
    ) +
    (
      (u[1] * u_scale[1] * u_scale[2] * dt + u_scale[2] * x[3]) *
      sin(u[2] * u_scale[2] * dt + x[4]) -
      u[1] * u_scale[1] * u_scale[2] * dt * sin(u[2] * u_scale[2] * dt + x[4]) +
      u_scale[2] *
      dt *
      (u[1] * u_scale[1] * u[2] * u_scale[2] * dt + u[2] * u_scale[2] * x[3]) *
      cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) +
    (
      2 *
      (u[2] * u_scale[2] * x[3] * sin(x[4]) + u[1] * u_scale[1] * cos(x[4]))
    ) / (u[2]^3 * u_scale[2]^2) - (x[3] * sin(x[4])) / (u[2]^2 * u_scale[2])
  fu[2, 1] =
    (
      u_scale[1] * sin(u[2] * u_scale[2] * dt + x[4]) -
      u_scale[1] * u[2] * u_scale[2] * dt * cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (u_scale[1] * sin(x[4])) / (u[2]^2 * u_scale[2]^2)
  fu[2, 2] =
    (
      (
        -u_scale[2] *
        dt *
        (
          (-u[1] * u_scale[1] * u[2] * u_scale[2] * dt) -
          u[2] * u_scale[2] * x[3]
        ) *
        sin(u[2] * u_scale[2] * dt + x[4])
      ) +
      ((-u[1] * u_scale[1] * u_scale[2] * dt) - u_scale[2] * x[3]) *
      cos(u[2] * u_scale[2] * dt + x[4]) +
      u[1] * u_scale[1] * u_scale[2] * dt * cos(u[2] * u_scale[2] * dt + x[4])
    ) / (u[2]^2 * u_scale[2]^2) -
    (
      2 * (
        u[1] * u_scale[1] * sin(u[2] * u_scale[2] * dt + x[4]) +
        (
          (-u[1] * u_scale[1] * u[2] * u_scale[2] * dt) -
          u[2] * u_scale[2] * x[3]
        ) * cos(u[2] * u_scale[2] * dt + x[4])
      )
    ) / (u[2]^3 * u_scale[2]^2) +
    (
      2 *
      (u[1] * u_scale[1] * sin(x[4]) - u[2] * u_scale[2] * x[3] * cos(x[4]))
    ) / (u[2]^3 * u_scale[2]^2) +
    (x[3] * cos(x[4])) / (u[2]^2 * u_scale[2])
  fu[3, 1] = u_scale[1] * dt
  fu[3, 2] = 0
  fu[4, 1] = 0
  fu[4, 2] = u_scale[2] * dt
  return fu
end
##$#############################################################################
##^# create the SCP dynamics matrix ############################################
function linearized_dynamics(x0, f, fx, fu, X_prev, U_prev)
  xdim, udim, N = size(fu)
  F = zeros(N * xdim, N * udim)
  for i = 1:N
    for j = 1:(i - 1)
      Fm1 = F[
        ((i - 2) * xdim + 1):((i - 1) * xdim),
        ((j - 1) * udim + 1):(j * udim),
      ]
      F[((i - 1) * xdim + 1):(i * xdim), ((j - 1) * udim + 1):(j * udim)] =
        fx[:, :, i] * Fm1
    end
    F[((i - 1) * xdim + 1):(i * xdim), ((i - 1) * udim + 1):(i * udim)] =
      fu[:, :, i]
  end
  f_ = zeros(xdim, N)
  f_[:, 1] = f[:, 1] - fu[:, :, 1] * U_prev[:, 1]
  for i = 2:N
    f_[:, i] =
      f[:, i] + fx[:, :, i] * (f_[:, i - 1] - X_prev[:, i]) -
      fu[:, :, i] * U_prev[:, i]
  end
  return F, reshape(f_, :)
end
##$#############################################################################

File: IntelligentDriving.jl/resources/python/dynamics.py
##^# library imports ###########################################################
import os, pdb, sys, time
import matplotlib.pyplot as plt, numpy as np, torch

##$#############################################################################
##^# dynamics ##################################################################
def car(x, u, p, pkg=torch):
    """
    unicycle car dynamics, 4 states, 2 actions
    x1: position x
    x2: position y
    x3: speed (local frame)
    x4: orientation angle

    u1: acceleration
    u2: turning speed (independent of velocity)
    """
    assert x.shape[-1] == 4 and u.shape[-1] == 2
    T, u_scale1, u_scale2 = p[..., 0], p[..., 1], p[..., 2]
    eps = 1e-6 * pkg.ones(())
    u1, u2 = u_scale1 * u[..., 0], -u_scale2 * u[..., 1]
    u1 = u1 + pkg.where(u1 >= 0.0, eps, -eps)
    u2 = u2 + pkg.where(u2 >= 0.0, eps, -eps)

    x0, y0, v0, th0 = x[..., 0], x[..., 1], x[..., 2], x[..., 3]
    xp = pkg.zeros(x.shape)
    xp1 = (
        x0
        + (
            u1 * pkg.sin(T * u1 + th0) * v0
            + (T * u1 * pkg.sin(T * u1 + th0) + pkg.cos(T * u1 + th0)) * u2
        )
        / u1 ** 2
        - (pkg.sin(th0) * u1 * v0 + pkg.cos(th0) * u2) / u1 ** 2
    )
    xp2 = (
        y0
        + (pkg.cos(th0) * u1 * v0 - pkg.sin(th0) * u2) / u1 ** 2
        - (
            u1 * pkg.cos(T * u1 + th0) * v0
            + (T * u1 * pkg.cos(T * u1 + th0) - pkg.sin(T * u1 + th0)) * u2
        )
        / u1 ** 2
    )
    xp3 = v0 + T * u1
    xp4 = th0 + T * u2
    xp = pkg.cat(
        [xp1[..., None], xp2[..., None], xp3[..., None], xp4[..., None]], -1
    )
    return xp

def hello():
    print("Hello World")


def f_fx_fu_fn(x, u, p):
    """
    This function accomodates Julia style batched inputs (xdim,) + bshape
    """
    xdim, udim, bshape = x.shape[0], u.shape[0], x.shape[1:]
    assert bshape == u.shape[1:] and bshape == p.shape[1:]

    x, u, p = [z.reshape((z.shape[0], -1)).swapaxes(-2, -1) for z in [x, u, p]]
    x, u, p = torch.as_tensor(x), torch.as_tensor(u), torch.as_tensor(p)
    x.requires_grad = True
    u.requires_grad = True
    f = car(x, u, p)
    gs_list = zip(
        *[
            torch.autograd.grad(
                torch.sum(f, tuple(range(f.ndim - 1)))[i],
                (x, u),
                retain_graph=(i - 1 < f.shape[-1]),
            )
            for i in range(f.shape[-1])
        ]
    )
    fx, fu = [torch.stack(gs, -2) for gs in gs_list]
    fx = (
        fx.reshape((-1, xdim * xdim))
        .transpose(-2, -1)
        .reshape((xdim, xdim) + bshape)
    )
    fu = (
        fu.reshape((-1, xdim * udim))
        .transpose(-2, -1)
        .reshape((xdim, udim) + bshape)
    )
    f = f.detach().transpose(-2, -1).reshape((xdim,) + bshape)
    return f.numpy(), fx.numpy(), fu.numpy()


##$#############################################################################
##^# linearization tools #######################################################
def dyn_mat(x0, f, fx, fu, X_prev, U_prev, pkg=torch):
    """
    construct the matrix and bias vector that gives from a local linearization
    vec(X) = Ft @ vec(U) + ft
    """
    bshape, (N, xdim), udim = fx.shape[:-3], fx.shape[-3:-1], fu.shape[-1]
    Fts = [[None for _ in range(N)] for _ in range(N)]
    Z_ = pkg.zeros(bshape + (xdim, udim))
    Fts = [[Z_ for _ in range(N)] for _ in range(N)]
    for i in range(N):
        for j in range(i):
            Fts[i][j] = fx[..., i, :, :] @ Fts[i - 1][j]
        Fts[i][i] = fu[..., i, :, :]
    Ft = pkg.cat(
        [pkg.cat([Fts[i][j] for i in range(N)], -2) for j in range(N)], -1
    )

    fts = [None for i in range(N)]
    f_ = f - bmv(fx, X_prev) - bmv(fu, U_prev)
    fts[0] = bmv(fx[..., 0, :, :], x0) + f_[..., 0, :]
    for i in range(1, N):
        fts[i] = bmv(fx[..., i, :, :], fts[i - 1]) + f_[..., i, :]
    ft = pkg.cat(fts, -1)
    return Ft, ft


##$#############################################################################
##^# tests #####################################################################
if __name__ == "__main__":
    bshape = (60,)
    X = np.random.randn(*((4,) + bshape))
    U = np.random.randn(*((2,) + bshape))
    p = 0.1 * np.ones((1,) + bshape)
    t = time.time()
    M = 10 ** 2
    for _ in range(M):
        f, fx, fu = f_fx_fu_fn(X, U, p)
    t = time.time() - t
    print("Time elapsed: %9.4e" % (t / M))
    print(f.shape)
    print(fx.shape)
    print(fu.shape)
    print()
    print()
    print()
    print(fx[:, :, 0])
    print(fu[:, :, 0])
##$#############################################################################
